contracts/gateway/RIFGateway.sol

```diff
@@ -56,14 +53,33 @@ contract RIFGateway is Ownable, SubscriptionReporter, IRIFGateway {
         return (_allServices, _providers);
     }
 
+    function _addProviderIfNotExists(address provider) internal {
+        if (_providerIndexes[provider] == 0) {
+            _providers.push(Provider({provider: provider, validated: false}));
+            _providerIndexes[provider] = _providers.length;
+        }
+    }
+
+    function _checkIfProviderIsAlreadyValidated(address provider)
+        internal
+        view
+    {
+        if (_providers[_providerIndexes[provider] - 1].validated) {
+            revert ProviderAlreadyValidated(provider);
+        }
+    }
+
     function requestValidation(address provider) external override {
-        if (_providerIndexes[provider] == 0) revert InvalidProvider(provider);
-        if (!_providers[_providerIndexes[provider] - 1].validated)
-            emit ValidationRequested(provider);
+        _addProviderIfNotExists(provider);
+        _checkIfProviderIsAlreadyValidated(provider);
+
+        emit ValidationRequested(provider);
     }
 
     function validateProvider(address provider) external override onlyOwner {
-        if (_providerIndexes[provider] == 0) revert InvalidProvider(provider);
+        _addProviderIfNotExists(provider);
```

Comment:
Per our validation flow not sure if this makes sense here. We should definitively check if the provider exists (it should as this functions will be call after `requestValidation`) before changing it's validated state to avoid an exception occurring but we must not add it by default.

What's your take on it?

Reply:
Yes, that’s exactly what I mentioned in the description, in the “TODO”.

but i think it needs to change a bit, in order to add an intermediate state “pending Validation”  so that we can identify if validation was already requested but not approved just yet. 

What do you think?

Reply:
As discussed adding states will increase complexity significantly. Your new commit addresses the issue of a non requested validation in case we hit a bug within the validation flow. LGTM.